proc/cut_test.go

```diff
@@ -3,76 +3,96 @@ package proc_test
-		{"foo", zeek.TypeString},
-		{"bar", zeek.TypeString},
-	})
-	r1, err = zson.NewRecordZeekStrings(fooBarDesc, "foo1", "bar1")
-	r2, err = zson.NewRecordZeekStrings(fooBarDesc, "foo2", "bar2")
-	r3, err = zson.NewRecordZeekStrings(fooBarDesc, "foo3", "bar3")
-	fooBarBatch := zson.NewArray([]*zson.Record{r1, r2, r3}, nano.MaxSpan)
+// Data sets for tests
+
+const fooOnly = `
+#0:record[foo:string]
+0:[foo1;]
+0:[foo2;]
+0:[foo3;]
+`
+
+const barOnly = `
+#1:record[bar:string]
+1:[bar1;]
+1:[bar2;]
+1:[bar3;]
+`
+
+const fooAndBar = `
+#0:record[foo:string,bar:string]
+0:[foo1;bar1;]
+0:[foo2;bar2;]
+0:[foo3;bar3;]
+`
 
+func TestCut(t *testing.T) {
 	// test "cut foo" on records that only have field foo
-	pt, err := proc.NewProcTestFromSource("cut foo", resolver, []zson.Batch{fooBatch})
-	require.NoError(t, err)
-	require.NoError(t, pt.Expect(fooBatch))
-	require.NoError(t, pt.ExpectEOS())
-	require.NoError(t, pt.Finish())
+	proc.TestOneProc(t, fooOnly, fooOnly, "cut foo")
 
 	// test "cut foo" on records that have fields foo and bar
-	pt, err = proc.NewProcTestFromSource("cut foo", resolver, []zson.Batch{fooBarBatch})
-	require.NoError(t, err)
-	require.NoError(t, pt.Expect(fooBatch))
-	require.NoError(t, pt.ExpectEOS())
-	require.NoError(t, pt.Finish())
+	proc.TestOneProc(t, fooAndBar, fooOnly, "cut foo")
 
 	// test "cut foo" on records that don't have field foo
-	pt, err = proc.NewProcTestFromSource("cut foo", resolver, []zson.Batch{barBatch})
-	require.NoError(t, err)
-	require.NoError(t, pt.ExpectEOS())
-	require.NoError(t, pt.ExpectWarning("Cut field foo not present in input"))
-	require.NoError(t, pt.Finish())
+	warning := "Cut field foo not present in input"
+	proc.TestOneProcWithWarnings(t, barOnly, "", []string{warning}, "cut foo")
 
 	// test "cut foo" on some fields with foo, some without
 	// Note there is no warning in this case since some of the input
 	// records have field "foo".
-	pt, err = proc.NewProcTestFromSource("cut foo", resolver, []zson.Batch{fooBatch, barBatch})
-	require.NoError(t, err)
-	require.NoError(t, pt.Expect(fooBatch))
-	require.NoError(t, pt.ExpectEOS())
-	require.NoError(t, pt.Finish())
+	proc.TestOneProc(t, fooOnly+barOnly, fooOnly, "cut foo")
 
 	// test cut on multiple fields.
-	pt, err = proc.NewProcTestFromSource("cut foo,bar", resolver, []zson.Batch{fooBarBatch})
-	require.NoError(t, err)
-	require.NoError(t, pt.Expect(fooBarBatch))
-	require.NoError(t, pt.ExpectEOS())
-	require.NoError(t, pt.Finish())
+	proc.TestOneProc(t, fooAndBar, fooAndBar, "cut foo,bar")
+}
+
+// Test that illegal cut operations fail at compile time with a
+// reasonable error message.
+func testNonAdjacentFields(t *testing.T, zql string) {
+	_, err := proc.CompileTestProc(zql, nil, nil)
+	require.Error(t, err, "cut with non-adjacent records failed")
+	require.Regexp(t, "All cut fields in record \\S+ must be adjacent", err.Error(), "error message for cutting non-adjacent fields is correct")
```

Comment:
Relying on the error text this way is a little fragile... I realize you couldn't just define a static error value and compare to that since you have the record name in the error message (👍 for that). However using the new fancy [wrapped errors](https://blog.golang.org/go1.13-errors), it should be possible to compare to an underlying static error  that is wrapped with the more informative message.

Reply:
I did this by just creating a new struct that implements the error interface, no wrapping involved.  I'm not sure what's idiomatic go, certainly open to further refinements here...